### PR TITLE
[WIP]Only execute codefolding preprocessor when requested

### PR DESCRIPTION
**Pull request by juhasch**
- [x] Only execute preprocessor when configuration variable is set
- [x] add test
- [x] add documentation

Addresses #792

### CODE REVIEW

**Comment by juhasch** [comment]
@jcb91 : I would like to move the `autoclass` API docs to a separate chapter `internals`, and instead only have the description on how to use the different exporters, pre- and preprocessors under `exporting` ?

**Comment by jcb91** [comment]
> I would like to move the `autoclass` API docs to a separate chapter `internals`, and instead only have the description on how to use the different exporters, pre- and preprocessors under `exporting`?

Sure, seems like a sensible move. I think we even already have an internals file :)

**Comment by juhasch** [comment]
Is there any deeper meaning why `exporting.rst` is `rst` and `internals.md` is `md` format ?
Otherwise I will convert the internals stuff to `rst`.

**Comment by jcb91** [comment]
No, I just kept to markdown for anything that didn't require rat-only
features, to make it easier to compare with the main (markdown) readme.
Convert away :)

File: docs/source/internals.rst [hunk 1]
```diff
@@ -0,0 +1,111 @@
+.. module:: jupyter_contrib_nbextensions.nbconvert_support
+
+Notebook extension structure
+============================
+
+The nbextensions are stored each as a separate subdirectory of
+`src/jupyter_contrib_nbextensions/nbextensions`
```
Comment (jcb91) [inline docs/source/internals.rst]: in rst, inline literals may need two backticks? I believe the notebook repo has somehow enabled use of single backticks, but I'm not sure how. The only reference to single backticks that I can find is [this](http://www.sphinx-doc.org/en/1.4.9/markup/inline.html), which seems to suggest that by default, single backticks have no special meaning

Comment (jcb91) [inline docs/source/internals.rst]: it seems the single backticks [give the default role](http://docutils.sourceforge.net/docs/ref/rst/roles.html), which [is configurable](http://docutils.sourceforge.net/docs/ref/rst/directives.html#default-role). Perhaps we should add a

```python
default_role = 'code'
```

to `conf.py`?

Comment (juhasch) [inline docs/source/internals.rst]: Good catch. Yes, I think we should set the default role.

**Comment by juhasch** [comment]
@jcb91 Any idea why one test fails ? I don't see how my changes would affect it.

**Comment by jcb91** [comment]
It's the 4.x notebook branch, and it's just failing to build notebook,
before even running our tests, so it's not really anything to do with your
changes, I think. Looks fine to merge to me!

**Pull request merged**